system/application/views/modules/dashboard/book_users.php

```diff
@@ -55,44 +55,88 @@ function remove_book_user(book_id, user_id) {
 				window.location.reload();
 			});
 		}
+
+		function debounce(func, wait, immediate) {
+	    let timeout;
+	    return function() {
+		    let context = this, args = arguments;
+		    const later = function() {
+			    timeout = null;
+			    if (!immediate) func.apply(context, args);
+		    };
+		    const callNow = immediate && !timeout;
+		    clearTimeout(timeout);
+		    timeout = setTimeout(later, wait);
+		    if (callNow) func.apply(context, args);
+	    };
+    };
+
 		function request_book_user(book_id) {
-				var $div = $('<div class="select_box"><h4 class="dialog_title">Add a user</h4>To connect a user to your book, first search for them by their full name.<br clear="both" /><br /><form><input class="generic_text_input" style="float:left;" type="text" name="fullname" value="Full name" /><input class="generic_button" style="float:left; margin-left:8px;" type="submit" value="Search" /><br clear="both" /></form><div class="results" style="padding-top:16px;padding-bottom:10px;"></div><a class="generic_button large" href="javascript:;" onclick="$(this).parent().remove();" style="float:right;font-size:larger;">Cancel</a></div>');
-				$div.find('input:first').on('focus', function() {if ($(this).val() == 'Full name') $(this).val('');});
-				$('body').append($div);
-				$div.find('form:first').on('submit', function() {
-					if ($div.find('input:first').val()=='Full name') return false;
-					$div.find('input:submit').attr("disabled", "disabled");
-					var fullname = this.fullname.value;
-					$.get('api/user_search', {fullname:fullname}, function(data) {
+		  let bookUserIDs = Array.prototype.map
+			  .call(document.querySelectorAll("[property='id']"), row => row.innerHTML);

```

Comment:
Slick use of `map()` on the results of `document.querySelectorAll()`. I'm going to have to remember that one!